File: pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/PartitionedBrokerConsumerStatsImpl.java
/**
 * Copyright 2016 Yahoo Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.yahoo.pulsar.client.impl;

import java.util.List;

import com.google.common.collect.Lists;
import com.yahoo.pulsar.client.api.BrokerConsumerStats;
import com.yahoo.pulsar.client.api.PulsarClientException;
import com.yahoo.pulsar.client.api.PulsarClientException.InvalidConfigurationException;
import com.yahoo.pulsar.client.api.SubscriptionType;

public class PartitionedBrokerConsumerStatsImpl implements BrokerConsumerStats {

    private String DELIMITER = ";";

    List<BrokerConsumerStats> brokerConsumerStatsList;

    PartitionedBrokerConsumerStatsImpl() {
        brokerConsumerStatsList = Lists.newArrayList();
    }

    public synchronized boolean isValid() {
        return brokerConsumerStatsList.stream().reduce(true, (accumulated, stats) -> accumulated & stats.isValid(),
                (accumulated1, accumulated2) -> accumulated1 & accumulated2);
    }

    public double getMsgRateOut() {
        return brokerConsumerStatsList.stream().mapToDouble(stats -> stats.getMsgRateOut()).sum();
    }

    public double getMsgThroughputOut() {
        return brokerConsumerStatsList.stream().mapToDouble(stats -> stats.getMsgThroughputOut()).sum();
    }

    public double getMsgRateRedeliver() {
        return brokerConsumerStatsList.stream().mapToDouble(stats -> stats.getMsgRateRedeliver()).sum();
    }

    public String getConsumerName() {
        return brokerConsumerStatsList.stream().reduce("",
                (accumulator, stats) -> stats.getConsumerName() + DELIMITER + accumulator,
                (acc1, acc2) -> acc1 + DELIMITER + acc2);
    }

    public long getAvailablePermits() {
        return brokerConsumerStatsList.stream().mapToLong(stats -> stats.getAvailablePermits()).sum();
    }

    public long getUnackedMessages() {
        return brokerConsumerStatsList.stream().mapToLong(stats -> stats.getUnackedMessages()).sum();
    }

    public boolean isBlockedConsumerOnUnackedMsgs() {
        return brokerConsumerStatsList.stream().reduce(true,
                (accumulated, stats) -> accumulated & stats.isBlockedConsumerOnUnackedMsgs(),
                (accumulated1, accumulated2) -> accumulated1 & accumulated2);
    }

    public String getAddress() {
        return brokerConsumerStatsList.stream().reduce("",
                (accumulator, stats) -> stats.getAddress() + DELIMITER + accumulator,
                (acc1, acc2) -> acc1 + DELIMITER + acc2);
    }

    public String getConnectedSince() {
        return brokerConsumerStatsList.stream().reduce("",
                (accumulator, stats) -> stats.getConnectedSince() + DELIMITER + accumulator,
                (acc1, acc2) -> acc1 + DELIMITER + acc2);
    }

    public SubscriptionType getSubscriptionType() {
        if (brokerConsumerStatsList.size() > 0) {
            return brokerConsumerStatsList.get(0).getSubscriptionType();
        }
        return SubscriptionType.Exclusive;
    }

    public double getMsgRateExpired() {
        return brokerConsumerStatsList.stream().mapToDouble(stats -> stats.getMsgRateExpired()).sum();
    }

    public long getMsgBacklog() {
        return brokerConsumerStatsList.stream().mapToLong(stats -> stats.getMsgBacklog()).sum();
    }

    public synchronized void add(BrokerConsumerStats stats) {
        brokerConsumerStatsList.add(stats);
    }

    public synchronized BrokerConsumerStats get(int partitionIndex) throws InvalidConfigurationException {
        int size = brokerConsumerStatsList.size();
        if (partitionIndex < 0 || partitionIndex >= size) {
            throw new PulsarClientException.InvalidConfigurationException("partitionIndex [" + partitionIndex
                    + "] needs to be positive and less than brokerConsumerStatsList.size() [" + size + "]");
        }
        return brokerConsumerStatsList.get(partitionIndex);
    }

    @Override
    public String toString() {
        return "PartitionedBrokerConsumerStatsImpl [" + brokerConsumerStatsList.stream().reduce("",
                (acc, stats) -> stats + "," + acc, (acc1, acc2) -> acc1 + "," + acc2) + "]";
    }
}

File: pulsar-client-cpp/lib/PartitionedBrokerConsumerStatsImpl.cc
/**
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */
#include <lib/PartitionedBrokerConsumerStatsImpl.h>
#include <boost/date_time/local_time/local_time.hpp>
#include <algorithm>
#include <numeric>

namespace pulsar {

const std::string PartitionedBrokerConsumerStatsImpl::DELIMITER = ";";

PartitionedBrokerConsumerStatsImpl::PartitionedBrokerConsumerStatsImpl(size_t size) {
    statsList_.resize(size);
}

bool PartitionedBrokerConsumerStatsImpl::isValid() const {
    bool isValid = true;
    for (int i = 0; i < statsList_.size(); i++) {
        isValid &= statsList_[i].isValid();
    }
    return isValid;
}

std::ostream& operator<<(std::ostream& os, const PartitionedBrokerConsumerStatsImpl& obj) {
    os << "\nPartitionedBrokerConsumerStatsImpl ["
       << "validTill_ = " << obj.isValid() << ", msgRateOut_ = " << obj.getMsgRateOut()
       << ", msgThroughputOut_ = " << obj.getMsgThroughputOut()
       << ", msgRateRedeliver_ = " << obj.getMsgRateRedeliver()
       << ", consumerName_ = " << obj.getConsumerName()
       << ", availablePermits_ = " << obj.getAvailablePermits()
       << ", unackedMessages_ = " << obj.getUnackedMessages()
       << ", blockedConsumerOnUnackedMsgs_ = " << obj.isBlockedConsumerOnUnackedMsgs()
       << ", address_ = " << obj.getAddress() << ", connectedSince_ = " << obj.getConnectedSince()
       << ", type_ = " << obj.getType() << ", msgRateExpired_ = " << obj.getMsgRateExpired()
       << ", msgBacklog_ = " << obj.getMsgBacklog() << "]";
    return os;
}

double PartitionedBrokerConsumerStatsImpl::getMsgRateOut() const {
    double sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getMsgRateOut();
    }
    return sum;
}

double PartitionedBrokerConsumerStatsImpl::getMsgThroughputOut() const {
    double sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getMsgThroughputOut();
    }
    return sum;
}

double PartitionedBrokerConsumerStatsImpl::getMsgRateRedeliver() const {
    double sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getMsgRateRedeliver();
    }
    return sum;
}

const std::string PartitionedBrokerConsumerStatsImpl::getConsumerName() const {
    std::string str;
    for (int i = 0; i < statsList_.size(); i++) {
        str += statsList_[i].getConsumerName() + DELIMITER;
    }
    return str;
}

uint64_t PartitionedBrokerConsumerStatsImpl::getAvailablePermits() const {
    uint64_t sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getAvailablePermits();
    }
    return sum;
}

uint64_t PartitionedBrokerConsumerStatsImpl::getUnackedMessages() const {
    uint64_t sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getUnackedMessages();
    }
    return sum;
}

bool PartitionedBrokerConsumerStatsImpl::isBlockedConsumerOnUnackedMsgs() const {
    if (statsList_.size() == 0) {
        return false;
    }

    bool isValid = true;
    for (int i = 0; i < statsList_.size(); i++) {
        isValid &= statsList_[i].isValid();
    }
    return isValid;
}

const std::string PartitionedBrokerConsumerStatsImpl::getAddress() const {
    std::string str;
    for (int i = 0; i < statsList_.size(); i++) {
        str += statsList_[i].getAddress() + DELIMITER;
    }
    return str;
}

const std::string PartitionedBrokerConsumerStatsImpl::getConnectedSince() const {
    std::string str;
    for (int i = 0; i < statsList_.size(); i++) {
        str += statsList_[i].getConnectedSince() + DELIMITER;
    }
    return str;
}

const ConsumerType PartitionedBrokerConsumerStatsImpl::getType() const {
    if (!statsList_.size()) {
        return ConsumerExclusive;
    }
    return statsList_[0].getType();
}

double PartitionedBrokerConsumerStatsImpl::getMsgRateExpired() const {
    double sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getMsgRateExpired();
    }
    return sum;
}

uint64_t PartitionedBrokerConsumerStatsImpl::getMsgBacklog() const {
    uint64_t sum = 0;
    for (int i = 0; i < statsList_.size(); i++) {
        sum += statsList_[i].getMsgBacklog();
    }
    return sum;
}

BrokerConsumerStats PartitionedBrokerConsumerStatsImpl::getBrokerConsumerStats(int index) {
    return statsList_[index];
}

void PartitionedBrokerConsumerStatsImpl::add(BrokerConsumerStats stats, int index) {
    statsList_[index] = stats;
}

void PartitionedBrokerConsumerStatsImpl::clear() { statsList_.clear(); }
}  // namespace pulsar
